### PR TITLE
Treat Kotlin nullable as non-required

### PR DESCRIPTION
**Pull request by rocketraman**
Modify the core MethodParameter class to return the Kotlin nullability
of a parameteter (cached for performance).

Where `isOptional` is used, also check for `isKotlinNullable` i.e.
values are not considered required if they are Kotlin nullables:
- spring-messaging: named value method arguments
- spring-web: named value method arguments
- spring-webmvc: request parts

This means that Kotlin client code no longer has to explicity specify
"required=false" for Kotlin nullables -- this information is inferred
automatically by the framework.

Issue: SPR-14165

### CODE REVIEW

**Comment by rocketraman** [comment]
Notes on submission / pending work:

* I put the kotlin-specific code into a separate module `spring-core-kotlin` to avoid adding the kotlin compiler to the `spring-core` module (inspired from `spring-beans-groovy`).
* I wasn't sure where to add tests for this. My first though was `MethodParameterTests`, but currently it contains very little in the way of tests. Guidance is appreciated.

Otherwise the implementation turned out to be quite simple and concise.

**Comment by sdeleuze** [comment]
Thanks for this PR, I only had a quick look but I would be in favor of not creating a new `spring-core-kotlin`, instead you could add kotlin dependency as an optional one. After that I will make a detailed review of your PR.

**Comment by rocketraman** [comment]
> Thanks for this PR, I only had a quick look but I would be in favor of not creating a new spring-core-kotlin, instead you could add kotlin dependency as an optional one. After that I will make a detailed review of your PR.

Done!

**Comment by rocketraman** [comment]
NOTE: Another place we could use this functionality is when `@Inject` or `@Autowired` is used, the Kotlin nullable information could be used to determine if the value is required or not.

**Comment by rocketraman** [comment]
@sdeleuze Ping. Looks like it needs another rebase, but would appreciate a review.

File: spring-core/src/main/java/org/springframework/core/MethodParameter.java [hunk 1]
```diff
@@ -329,6 +338,19 @@ public MethodParameter nestedIfOptional() {
 		return (isOptional() ? nested() : this);
 	}
 
+	/**
+	 * Return whether this method parameter is declared as a Kotlin
+	 * nullable.
+	 * @since 4.4
```
Comment (sdeleuze) [inline spring-core/src/main/java/org/springframework/core/MethodParameter.java]: Let's target 5.0 for this feature

File: spring-core/src/main/java/org/springframework/core/MethodParameter.java [hunk 2]
```diff
@@ -329,6 +338,19 @@ public MethodParameter nestedIfOptional() {
...
+	 */
+	public boolean isKotlinNullable() {
```
Comment (sdeleuze) [inline spring-core/src/main/java/org/springframework/core/MethodParameter.java]: Maybe `isNullable()`would be less specific and let the door opened to supporting this is other languages or with specific annotations

File: spring-core/src/main/java/org/springframework/core/MethodParameter.java [hunk 3]
```diff
@@ -50,6 +51,12 @@
  */
 public class MethodParameter {
 
+	private static final boolean kotlinPresent;
```
Comment (sdeleuze) [inline spring-core/src/main/java/org/springframework/core/MethodParameter.java]: Let's encapsulate that in a KotlinUtils class

File: spring-core/src/main/java/org/springframework/core/MethodParameter.java [hunk 4]
```diff
@@ -67,6 +74,8 @@
 
 	private volatile Class<?> parameterType;
 
+	private volatile Boolean isKotlinNullable;
```
Comment (sdeleuze) [inline spring-core/src/main/java/org/springframework/core/MethodParameter.java]: I think I would try initially without caching the result, unless you have specific concern about the performances

File: spring-core/src/main/kotlin/org/springframework/core/MethodParameterKotlin.kt [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+import kotlin.reflect.jvm.kotlinFunction
+
+/**
+ * Helper internal class that provides Kotlin-specific extensions to [MethodParameter].
+ *
+ * @author Raman Gupta
+ * @since 4.4
+ * @see MethodParameter
+ */
+object MethodParameterKotlin {
```
Comment (sdeleuze) [inline spring-core/src/main/kotlin/org/springframework/core/MethodParameterKotlin.kt]: Could we turn this into a regular java `KotlinUtils` class with maybe an additional method to know if Kotlin is present? I think I would like to restrict Kotlin code to only test classes.

**Comment by sdeleuze** [comment]
Sorry for the delay and thanks for your updated PR. I have added comments on the relevant lines of your commit.

Could you also rebase that on top of the new master (Spring Framework 5.0) and add some Kotlin unit tests to check that works?

Good idea to support that also for `@Inject`, feel free to add that to your PR and update the title of this JIRA issue to make is less specific.

**Comment by rocketraman** [comment]
> Good idea to support that also for @Inject, feel free to add that to your PR and update the title of this JIRA issue to make is less specific.

It looks more complicated to do this [1], so I'd prefer to have this pull request reviewed as-is for now, just to keep things as simple as possible. We can add the `@Inject` functionality in a subsequent pull.

> Sorry for the delay and thanks for your updated PR. I have added comments on the relevant lines of your commit.
> 
> Could you also rebase that on top of the new master (Spring Framework 5.0) and add some Kotlin unit tests to check that works?

I have updated the pull with all of your comments taken into account. Thanks!

[1] It looks like a variation of `RequiredAnnotationBeanPostProcessor` is needed.

**Comment by rocketraman** [comment]
Ping @sdeleuze for review.

**Comment by sdeleuze** [comment]
Superseeded by #1242

**Pull request closed without merging**